staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go

```diff
@@ -594,10 +600,28 @@ func patchResource(
 				}
 				originalObjJS, originalPatchedObjJS = originalJS, patchedJS
 			case types.StrategicMergePatchType:
-				originalMap, patchMap, err := strategicPatchObject(codec, currentObject, patchJS, objToUpdate, versionedObj)
+				// Since the patch is applied on versioned objects, we need to convert the
+				// current object to versioned representation first.
+				gvks, _, err := typer.ObjectKinds(versionedObj)
 				if err != nil {
 					return nil, err
 				}
+				gvk := gvks[0]
+				currentVersionedObject, err := convertor.ConvertToVersion(currentObject, gvk.GroupVersion())
+				if err != nil {
+					return nil, err
+				}
+				versionedObjToUpdate, err := creater.New(gvk)
+				if err != nil {
+					return nil, err
+				}
+				originalMap, patchMap, err := strategicPatchObject(codec, currentVersionedObject, patchJS, versionedObjToUpdate, versionedObj)
```

Comment:
Even without unsafe conversion, the conversions should be pretty much the same what we have now.
Basically, the input/output objects here are unversioned. And the codec we are using is the "versioning" codec, which is also doing the conversion (the safe one) + marshaling/unmarshaling.

So if we change to unsafe (which unless I'm completely missing something we can do), this should be even better (+ the savings from more performant conversion to unstructured).